hibernate-core/src/main/java/org/hibernate/type/TypeHelper.java

```diff
@@ -191,7 +191,7 @@ public static Object[] replaceAssociations(
 			}
 			else {
 				final Type type = types[i];
-				if ( type.isComponentType() ) {
+				if ( type.isComponentType() && !type.isAnyType() ) {
```

Comment:
I think a better solution would be to move the block:

```java
else if ( !type.isAssociationType() ) {
    copied[i] = target[i];
}
```

to before `if ( type.isComponentType() ) {`.

Reply:
but in this way in case of component types the `!type.isAssociationType()` block is always executed

Reply:
Excuse me, I said that wrong, I mean:

```java
final Type type = types[i];
if ( type.isAssociationType() ) {
	copied[i] = types[i].replace( currentOriginal, target[i], session, owner, copyCache, foreignKeyDirection );
}
else if ( type.isComponentType() ) {
	final CompositeType compositeType = (CompositeType) type;
	if ( target[i] != null ) {
		// need to extract the component values and check for subtype replacements...
		final Object[] objects = replaceCompositeAssociations(
				session,
				copyCache,
				foreignKeyDirection,
				target[i],
				currentOriginal,
				compositeType
		);
		target[i] = compositeType.replacePropertyValues( target[i], objects, session );
	}
	copied[i] = target[i];
}
else {
	copied[i] = target[i];
}
```

Reply:
Or even better:

```java
final Type type = types[i];
if ( type.isAssociationType() ) {
    copied[i] = types[i].replace( currentOriginal, target[i], session, owner, copyCache, foreignKeyDirection );
}
else {
    if ( type.isComponentType() ) {
	final CompositeType compositeType = (CompositeType) type;
	if ( target[i] != null ) {
		// need to extract the component values and check for subtype replacements...
		final Object[] objects = replaceCompositeAssociations(
				session,
				copyCache,
				foreignKeyDirection,
				target[i],
				currentOriginal,
				compositeType
		);
		target[i] = compositeType.replacePropertyValues( target[i], objects, session );
	}
    }
    copied[i] = target[i];
}
```

Reply:
applied your suggestion, running tests locally.

Reply:
Thanks @gavinking, applied your suggestion to the PR

Reply:
Awesome, cheers!